### PR TITLE
feat: incentive accumulator migration

### PR DESCRIPTION
**Pull request by p0mvn**
Closes: #XXX

## What is the purpose of the change

This PR introduces incentive accumulator migration to address the truncation issue with pool 1423.

In the v23 upgrade, we introduce a scaling factor to avoid truncating the per-unit-of-liquidity incentive accumulator. See: https://github.com/osmosis-labs/osmosis/pull/7409

The goal is to have all the newly created pools utilize this accumulator. At the same time, we want to migrate the old pools that are currently affected (e.g. pool 1423). We also want to migrate only the affected pools to limit migration risk.

To achieve all of the above, this PR:
- Introduces a pool ID threshold snapshotted during the upgrade to state. After this threshold, all newly created pools utilize the scaling factor.
- A hard-coded map of pool IDs that are below the threshold but migrated to the scaling factor logic

In the core module logic, we dynamically select the scaling factor based on the pool ID to apply to the accumulator.

Given successful migration, we will migrate the remaining pools in a future upgrade.

### Testnet

Pool ID threshold is selected dynamically based on either mainnet or testnet state.

For any environment other than `osmosis-1`, we avoid performing pool migration to avoid dealing with complexity. This is. fine as we don't care about other environments potentially truncating with older pools.

## Testing and Verifying

- [ ] Test core helpers
- [ ] Test upgrade handler
- [ ] Test genesis

## Documentation and Release Note

  - [ ] Does this pull request introduce a new feature or user-facing behavior changes?
  - [ ] Changelog entry added to `Unreleased` section of `CHANGELOG.md`?

Where is the change documented? 
  - [ ] Specification (`x/{module}/README.md`)
  - [ ] Osmosis documentation site
  - [ ] Code comments?
  - [ ] N/A

### CODE REVIEW

File: x/concentrated-liquidity/types/migration.go [hunk 1]
```diff
@@ -0,0 +1,5 @@
+package types
+
+var MigratedIncentiveAccumulatorPoolIDs = map[uint64]struct{}{
+	1423: {},
+}
```
Comment (p0mvn) [inline x/concentrated-liquidity/types/migration.go:5]: Note: will extend the pool IDs separately

File: osmoutils/accum/accum.go [hunk 2]
```diff
@@ -65,6 +65,19 @@ func MakeAccumulatorWithValueAndShare(accumStore store.KVStore, accumName string
...
 }
 
+// OverwriteAccumulatorUnsafe overwrites the accumulator with the given name in accumStore.
+// Use with caution as this method is only meant for use in migrations.
+func OverwriteAccumulatorUnsafe(accumStore store.KVStore, accumName string, accumValue sdk.DecCoins, totalShares osmomath.Dec) error {
+	if !accumStore.Has(formatAccumPrefixKey(accumName)) {
+		return errors.New("Accumulator with given name does not exist in store")
+	}
+
+	newAccum := AccumulatorObject{accumStore, accumName, accumValue, totalShares}
```
Comment (ValarDragon) [inline osmoutils/accum/accum.go:75]: Is there any sanity check we can/should do wrt totalShares here? Or not really?

Comment (p0mvn) [reply osmoutils/accum/accum.go:75]: We could check if `accumValue` is non-negative or `totalShares` are not-zero but I don't think that the benefit of these checks would be greater than the added complexity.

IMO, keeping it as is fine. However, happy to implement checks if there are any specific suggestions

File: x/concentrated-liquidity/incentives.go [hunk 3]
```diff
@@ -1032,3 +1045,50 @@ func emitAccumulatorUpdateTelemetry(ctx sdk.Context, truncatedPlaceholder, emitt
...
+// It returns perUnitLiqScalingFactor if the pool is migrated or if the pool ID is greater than the migration threshold.
+// It returns oneDecScalingFactor otherwise.
+func (k Keeper) getIncentiveScalingFactorForPool(ctx sdk.Context, poolID uint64) (osmomath.Dec, error) {
+	migrationThreshold, err := k.GetIncentivePoolIDMigrationThreshold(ctx)
+	if err != nil {
+		return osmomath.Dec{}, err
+	}
+
+	// If the given pool ID is greater than the migration threshold, we return the perUnitLiqScalingFactor.
+	if poolID > migrationThreshold {
```
Comment (ValarDragon) [inline x/concentrated-liquidity/incentives.go:1059]: sanity checking we got > vs >= right?

File: x/concentrated-liquidity/pool.go [hunk 4]
```diff
@@ -418,3 +420,97 @@ func (k Keeper) GetUserUnbondingPositions(ctx sdk.Context, address sdk.AccAddres
...
+	}
+
+	// Get all position IDs for the pool.
+	positionIDs, err := k.getPositionIDsByPoolID(ctx, poolId)
+	if err != nil {
+		return err
+	}
+
+	// For each uptime accimulator, multiply the value by the per-unit liquidity scaling factor
+	// and overwrite the accumulator with the new value.
```
Comment (ValarDragon) [inline x/concentrated-liquidity/pool.go:464]: oh oof, have we measured how slow this is?

Comment (p0mvn) [reply x/concentrated-liquidity/pool.go:464]: Currently working on unblocking testnet. Will get back with the results

File: x/concentrated-liquidity/pool.go [hunk 5]
```diff
@@ -418,3 +420,97 @@ func (k Keeper) GetUserUnbondingPositions(ctx sdk.Context, address sdk.AccAddres
...
+
+			positionSnapshot, err := uptimeAccum.GetPosition(string(positionPrefix))
+			if err != nil {
+				return err
+			}
+
+			positionSnapshotValue := positionSnapshot.GetAccumValuePerShare()
+
+			// Multiply the value by the per-unit liquidity scaling factor
+			newValue := positionSnapshotValue.MulDecTruncate(perUnitLiqScalingFactor)
```
Comment (ValarDragon) [inline x/concentrated-liquidity/pool.go:488]: Is there a mutative method for this?

File: x/concentrated-liquidity/incentives.go (hunk 3 above)
Comment (p0mvn) [reply x/concentrated-liquidity/incentives.go:1059]: It is set to `lastPoolID`
https://github.com/osmosis-labs/osmosis/blob/68020b18694660308e1fd6e3ad9c505ca031498a/app/upgrades/v23/upgrades.go#L36-L40

File: x/concentrated-liquidity/pool.go [hunk 6]
```diff
@@ -418,3 +420,97 @@ func (k Keeper) GetUserUnbondingPositions(ctx sdk.Context, address sdk.AccAddres
...
+	}
+
+	// Retrieve all ticks for the pool
+	ticks, err := k.GetAllInitializedTicksForPool(ctx, poolId)
+	if err != nil {
+		return err
+	}
+
+	// For each tick, multiply the value of the uptime accumulator tracker.
+	for _, tick := range ticks {
```
Comment (ValarDragon) [inline x/concentrated-liquidity/pool.go:504]: yikes didn't realize this was for every tick as well

**Review by ValarDragon — APPROVED** [review]
Logic looks good to me, but im really worried about migration times. Should wait until we've tested it out ofc, but I suspect we may need to do something more lazy here.

Or at minimum split it up so we only migrate one pool at a time. (or maybe theres a clever way to only have to migrate nearby ticks immediately or smth)

File: x/concentrated-liquidity/pool.go [hunk 7]
```diff
@@ -418,3 +420,97 @@ func (k Keeper) GetUserUnbondingPositions(ctx sdk.Context, address sdk.AccAddres
 	}
 	return userPositionsWithPeriodLocks, nil
 }
+
+// getPositionIDsByPoolID returns all position IDs for a given pool ID.
+func (k Keeper) getPositionIDsByPoolID(ctx sdk.Context, poolID uint64) ([]uint64, error) {
```
Comment (czarcas7ic) [inline x/concentrated-liquidity/pool.go:425]: seems like a useful query to expose in the future.

Comment (p0mvn) [reply x/concentrated-liquidity/pool.go:425]: WIP in separate PR

**Review by czarcas7ic — APPROVED** [review]
Well done, logic appears sound to me. 

Of course need to benchmark to ensure no issues with duration of migration, but other than that LGTM.

File: x/concentrated-liquidity/incentives.go (hunk 3 above)
Comment (AlpinYukseloglu) [reply x/concentrated-liquidity/incentives.go:1059]: Looks right to me too

**Review by AlpinYukseloglu — APPROVED** [review]
LGTM, nice work. Iterating over ticks for migration seems unavoidable, so we'll have to see how migration time looks on benchmarks.

Since there's no long term loss to delaying the migration, we could also potentially split it up (would need to discuss what a non-hacky way to do this would be)

File: x/concentrated-liquidity/pool.go (hunk 4 above)
Comment (p0mvn) [reply x/concentrated-liquidity/pool.go:469]: Discussed offline

File: x/concentrated-liquidity/pool.go (hunk 5 above)
Comment (p0mvn) [reply x/concentrated-liquidity/pool.go:493]: Doesn't look like there is. made issue: https://github.com/cosmos/cosmos-sdk/issues/19389

File: x/concentrated-liquidity/incentives.go [hunk 8]
```diff
@@ -794,7 +807,7 @@ func (k Keeper) prepareClaimAllIncentivesForPosition(ctx sdk.Context, positionId
 			// We always truncate down in the pool's favor.
 			collectedIncentivesForUptime := sdk.NewCoins()
 			for _, incentiveCoin := range collectedIncentivesForUptimeScaled {
-				incentiveCoin.Amount = incentiveCoin.Amount.ToLegacyDec().QuoTruncateMut(perUnitLiqScalingFactor).TruncateInt()
+				incentiveCoin.Amount = incentiveCoin.Amount.ToLegacyDec().QuoTruncateMut(incentiveScalingFactor).TruncateInt()
```
[the remark this answers is not in the document]
Comment (p0mvn) [reply x/concentrated-liquidity/incentives.go]: addressed: [56b8146](https://github.com/osmosis-labs/osmosis/pull/7416/commits/56b8146697e429f4e461b2fdcff27d119f695b94)

**Pull request merged**